  # Challenge

  ## Identifying Outliers using Standard Deviation

In [91]:
# initial imports
import pandas as pd
import numpy as np
import random
from sqlalchemy import create_engine
from numpy import mean
from numpy import std
from numpy import percentile
import plotly.express as px


In [92]:
# create a connection to the database
engine = create_engine("postgresql://postgres:Istay@10314@localhost:5432/fraud_detection")

# loading data from the database

def execute_query(query):

    transaction_df = pd.read_sql(sql=query, con=engine, index_col='date', parse_dates='date')

    return transaction_df



In [93]:
# loading data of daily transactions from jan to jun 2018 for card holder 25
query = f'SELECT a.id, a.name, b.card, c.date, c.amount, e.name as "category" \
        FROM public.card_holder a, public.credit_card b, public.transaction c, public.merchant d, public.merchant_category e \
        WHERE a.id = b.id_card_holder AND b.card=c.card AND c.id_merchant=d.id AND d.id_merchant_category=e.id'

transaction_df = execute_query(query)
transaction_df.head()

id               name              card  amount  \
date                                                                   
2018-01-01 21:35:10  13        John Martin  3561954487988605    6.22   
2018-01-01 21:43:12  13        John Martin  5135837688671496    3.83   
2018-01-01 22:41:21  10  Matthew Gutierrez   213193946980303    9.61   
2018-01-01 23:13:30   4     Danielle Green  4263694062533017   19.03   
2018-01-01 23:15:10  18      Malik Carlson     4498002758300    2.95   

                       category  
date                             
2018-01-01 21:35:10  food truck  
2018-01-01 21:43:12         bar  
2018-01-01 22:41:21  food truck  
2018-01-01 23:13:30         pub  
2018-01-01 23:15:10  restaurant

In [94]:
# code a function to identify outliers based on standard deviation
# calculate summary statistics
data_mean, data_std = mean(transaction_df['amount']), std(transaction_df['amount'])

# identify outliers
cut_off = data_std * 3

lower, upper = data_mean - cut_off, data_mean + cut_off

# identify outliers
outliers = [x for x in transaction_df['amount'] if x < lower or x > upper]

print('Identified outliers: %d' % len(outliers))

# remove outliers
outliers_removed = [x for x in transaction_df['amount'] if x >= lower and x <= upper]

print('Non-outlier observations: %d' % len(outliers_removed))

transaction_df['outlier'] = (transaction_df['amount'] > upper) | (transaction_df['amount'] < lower)

outlier = transaction_df[transaction_df['outlier']==True]
outlier

Identified outliers: 77
Non-outlier observations: 3423


id              name                 card  amount  \
date                                                                     
2018-01-02 23:27:46  12       Megan Price         501879657465  1031.0   
2018-01-04 03:05:18   7       Sean Taylor     3516952396080247  1685.0   
2018-01-08 02:34:32   6    Beth Hernandez     3581345943543942  1029.0   
2018-01-22 08:07:03  16     Crystal Clark     5570600642865857  1131.0   
2018-01-23 06:29:37  12       Megan Price         501879657465  1678.0   
2018-01-24 13:17:19   1    Robert Johnson  4761049645711555811  1691.0   
2018-01-30 18:31:00  25   Nancy Contreras        4319653513507  1177.0   
2018-02-17 01:27:19  16     Crystal Clark     5570600642865857  1430.0   
2018-02-19 16:00:43   7       Sean Taylor     3516952396080247  1072.0   
2018-02-19 22:48:25  18     Malik Carlson      344119623920892  1839.0   
2018-02-27 15:27:32   6    Beth Hernandez     3581345943543942  1145.0   
2018-03-01 21:29:05   3  Elizabeth Sawyer       30078299053512  1119.0   
2018-03-04 15:50:53   9      Laurie Gibbs       30181963913340  1534.0   
2018-03-05 08:26:08  16     Crystal Clark     5570600642865857  1617.0   
2018-03-06 07:18:09  25   Nancy Contreras        4319653513507  1334.0   
2018-03-12 00:44:01  12       Megan Price         501879657465  1530.0   
2018-03-20 10:19:25  12       Megan Price         501879657465   852.0   
2018-03-20 13:05:54  24  Stephanie Dalton       30142966699187  1011.0   
2018-03-26 07:41:59   9      Laurie Gibbs       30181963913340  1009.0   
2018-04-03 03:23:37  18     Malik Carlson      344119623920892  1077.0   
2018-04-08 06:03:50  25   Nancy Contreras        4319653513507  1063.0   
2018-04-18 23:23:29   7       Sean Taylor     3516952396080247  1086.0   
2018-04-21 19:41:51   6    Beth Hernandez     3581345943543942  2108.0   
2018-05-08 13:21:01  24  Stephanie Dalton       30142966699187  1901.0   
2018-05-13 06:31:20  25   Nancy Contreras        4319653513507  1046.0   
2018-05-29 02:55:08  16     Crystal Clark     5570600642865857  1203.0   
2018-06-03 20:02:28  18     Malik Carlson      344119623920892  1814.0   
2018-06-04 03:46:15  25   Nancy Contreras        4319653513507  1162.0   
2018-06-06 21:50:17  25   Nancy Contreras        4319653513507   749.0   
2018-06-10 04:54:27   9      Laurie Gibbs       30181963913340  1795.0   
...                  ..               ...                  ...     ...   
2018-09-04 01:35:39   1    Robert Johnson  4761049645711555811  1790.0   
2018-09-06 08:28:55   1    Robert Johnson  4761049645711555811  1017.0   
2018-09-06 21:55:02   1    Robert Johnson  4761049645711555811  1056.0   
2018-09-10 22:49:41  18     Malik Carlson      344119623920892  1176.0   
2018-09-11 15:16:47   6    Beth Hernandez     3581345943543942  1856.0   
2018-09-23 19:20:23  12       Megan Price         501879657465  1075.0   
2018-09-25 23:23:21   9      Laurie Gibbs       30181963913340  1095.0   
2018-09-26 08:48:40   1    Robert Johnson  4761049645711555811  1060.0   
2018-10-07 14:40:34   3  Elizabeth Sawyer       30078299053512   757.0   
2018-10-07 18:29:20   9      Laurie Gibbs       30181963913340  1179.0   
2018-10-19 01:07:37   3  Elizabeth Sawyer       30078299053512  1053.0   
2018-11-13 17:07:25  16     Crystal Clark     5570600642865857  1911.0   
2018-11-17 05:30:43  18     Malik Carlson      344119623920892  1769.0   
2018-11-20 05:24:28   3  Elizabeth Sawyer       30078299053512  1054.0   
2018-11-25 20:44:07  12       Megan Price         501879657465  1123.0   
2018-11-27 15:36:05  12       Megan Price         501879657465  1802.0   
2018-11-27 17:20:29   6    Beth Hernandez     3581345943543942  1279.0   
2018-11-27 17:27:34   1    Robert Johnson  4761049645711555811  1660.0   
2018-12-03 02:38:52  16     Crystal Clark     5570600642865857  1014.0   
2018-12-07 07:22:03   1    Robert Johnson  4761049645711555811  1894.0   
2018-12-13 12:09:58  18     Malik Carlson      344119623920892  1154.0   
2018-12-13 15:51:59   7  

In [96]:
# find anomalous transactions for 3 random card holders
import datetime
start_time = datetime.time(7,0,0)
end_time = datetime.time(9,0,0)

anomalous_transactions = outlier.between_time(start_time, end_time).sort_values('amount', ascending=False)

px.scatter(anomalous_transactions, x='name', y='amount', color='category', title='Anomalous Transactions')

  ## Identifying Outliers Using Interquartile Range

In [86]:
# code a function to identify outliers based on interquartile range
# calculate interquartile range
q25, q75 = percentile(transaction_df['amount'], 25), percentile(transaction_df['amount'], 75)
iqr = q75 - q25
print('Percentiles: 25th=%.3f, 75th=%.3f, IQR=%.3f' % (q25, q75, iqr))
# calculate the outlier cutoff
cut_off = iqr * 1.5
lower, upper = q25 - cut_off, q75 + cut_off
# identify outliers
outliers_2 = [x for x in transaction_df['amount'] if x < lower or x > upper]
print('Identified outliers: %d' % len(outliers_2))
# remove outliers
outliers_removed_2 = [x for x in transaction_df['amount'] if x >= lower and x <= upper]
print('Non-outlier observations: %d' % len(outliers_removed_2))

transaction_df['outlier'] = (transaction_df['amount'] > upper) | (transaction_df['amount'] < lower)

outlier_2 = transaction_df[transaction_df['outlier']==True]
outlier_2


Percentiles: 25th=3.735, 75th=14.648, IQR=10.913
Identified outliers: 110
Non-outlier observations: 3390


id              name                 card  amount  \
date                                                                     
2018-01-02 23:27:46  12       Megan Price         501879657465  1031.0   
2018-01-04 03:05:18   7       Sean Taylor     3516952396080247  1685.0   
2018-01-07 01:10:54  18     Malik Carlson      344119623920892   175.0   
2018-01-08 02:34:32   6    Beth Hernandez     3581345943543942  1029.0   
2018-01-08 11:15:36  18     Malik Carlson      344119623920892   333.0   
2018-01-11 13:20:31  16     Crystal Clark     5570600642865857   229.0   
2018-01-22 08:07:03  16     Crystal Clark     5570600642865857  1131.0   
2018-01-23 06:29:37  12       Megan Price         501879657465  1678.0   
2018-01-24 13:17:19   1    Robert Johnson  4761049645711555811  1691.0   
2018-01-30 18:31:00  25   Nancy Contreras        4319653513507  1177.0   
2018-02-09 11:38:37   7       Sean Taylor     3516952396080247   445.0   
2018-02-17 01:27:19  16     Crystal Clark     5570600642865857  1430.0   
2018-02-19 16:00:43   7       Sean Taylor     3516952396080247  1072.0   
2018-02-19 22:48:25  18     Malik Carlson      344119623920892  1839.0   
2018-02-27 15:27:32   6    Beth Hernandez     3581345943543942  1145.0   
2018-03-01 21:29:05   3  Elizabeth Sawyer       30078299053512  1119.0   
2018-03-04 15:50:53   9      Laurie Gibbs       30181963913340  1534.0   
2018-03-05 08:26:08  16     Crystal Clark     5570600642865857  1617.0   
2018-03-06 07:18:09  25   Nancy Contreras        4319653513507  1334.0   
2018-03-09 04:51:38   6    Beth Hernandez     3581345943543942   389.0   
2018-03-12 00:44:01  12       Megan Price         501879657465  1530.0   
2018-03-20 10:19:25  12       Megan Price         501879657465   852.0   
2018-03-20 13:05:54  24  Stephanie Dalton       30142966699187  1011.0   
2018-03-26 07:41:59   9      Laurie Gibbs       30181963913340  1009.0   
2018-04-01 07:17:21  25   Nancy Contreras        4319653513507   100.0   
2018-04-03 03:23:37  18     Malik Carlson      344119623920892  1077.0   
2018-04-08 06:03:50  25   Nancy Contreras        4319653513507  1063.0   
2018-04-09 10:24:32   1    Robert Johnson  4761049645711555811   283.0   
2018-04-09 18:28:25  25   Nancy Contreras        4319653513507   269.0   
2018-04-10 06:08:01   7       Sean Taylor     3516952396080247   543.0   
...                  ..               ...                  ...     ...   
2018-10-11 23:29:33   3  Elizabeth Sawyer       30078299053512   206.0   
2018-10-16 13:27:33   1    Robert Johnson  4761049645711555811   484.0   
2018-10-19 01:07:37   3  Elizabeth Sawyer       30078299053512  1053.0   
2018-10-19 12:32:37  16     Crystal Clark     5570600642865857   178.0   
2018-10-23 22:47:13  16     Crystal Clark     5570600642865857   393.0   
2018-10-28 02:12:58  25   Nancy Contreras        4319653513507   137.0   
2018-11-13 05:58:47  24  Stephanie Dalton       30142966699187   466.0   
2018-11-13 17:07:25  16     Crystal Clark     5570600642865857  1911.0   
2018-11-17 05:30:43  18     Malik Carlson      344119623920892  1769.0   
2018-11-20 05:24:28   3  Elizabeth Sawyer       30078299053512  1054.0   
2018-11-23 09:08:05  12       Megan Price         501879657465   233.0   
2018-11-25 20:44:07  12       Megan Price         501879657465  1123.0   
2018-11-26 20:54:39   1    Robert Johnson  4761049645711555811   267.0   
2018-11-27 15:36:05  12       Megan Price         501879657465  1802.0   
2018-11-27 17:20:29   6    Beth Hernandez     3581345943543942  1279.0   
2018-11-27 17:27:34   1    Robert Johnson  4761049645711555811  1660.0   
2018-12-03 02:38:52  16     Crystal Clark     5570600642865857  1014.0   
2018-12-05 19:24:27   9      Laurie Gibbs       30181963913340    57.0   
2018-12-07 07:22:03   1    Robert Johnson  4761049645711555811  1894.0   
2018-12-13 12:09:58  18     Malik Carlson      344119623920892  1154.0   
2018-12-13 15:51:59   7       Sean Taylor     3516952396080247  2249.0   
2018-12-14 08:51:41  12  

In [89]:
# find anomalous transactions for 3 random card holders
anomalous_transactions2 = outlier_2.between_time(start_time, end_time).sort_values('amount', ascending=False)
anomalous_transactions2



id             name                 card  amount  \
date                                                                    
2018-12-07 07:22:03   1   Robert Johnson  4761049645711555811  1894.0   
2018-03-05 08:26:08  16    Crystal Clark     5570600642865857  1617.0   
2018-03-06 07:18:09  25  Nancy Contreras        4319653513507  1334.0   
2018-01-22 08:07:03  16    Crystal Clark     5570600642865857  1131.0   
2018-09-26 08:48:40   1   Robert Johnson  4761049645711555811  1060.0   
2018-09-06 08:28:55   1   Robert Johnson  4761049645711555811  1017.0   
2018-03-26 07:41:59   9     Laurie Gibbs       30181963913340  1009.0   
2018-12-14 08:51:41  12      Megan Price         501879657465   748.0   
2018-04-01 07:17:21  25  Nancy Contreras        4319653513507   100.0   

                        category  outlier  
date                                       
2018-12-07 07:22:03          bar     True  
2018-03-05 08:26:08          bar     True  
2018-03-06 07:18:09          bar     True  
2018-01-22 08:07:03   restaurant     True  
2018-09-26 08:48:40   restaurant     True  
2018-09-06 08:28:55          bar     True  
2018-03-26 07:41:59  coffee shop     True  
2018-12-14 08:51:41          pub     True  
2018-04-01 07:17:21  coffee shop     True

In [95]:
px.scatter(anomalous_transactions2, x='name', y='amount', color='category', title="Early Hour Transactions")

For Outlier calculation using standard deviation, results in 77 records whereas using Interquartile range results in 110 records. There seems to be fraudulent transactions in Bar category wherein amount spent between 7-9 AM in the Bar